larray/random.py

```diff
@@ -446,16 +447,77 @@ def choice(choices=None, axes=None, replace=True, p=None, meta=None):
     >>> la.random.choice(['hello', 'world', '!'], 3, replace=False, p=[0.1, 0.6, 0.3])  # doctest: +SKIP
     {0}*      0  1      2
           world  !  hello
+
+    Using an N-dimensional array as probabilities:
+
+    >>> proba = LArray([[0.15, 0.25, 0.10],
+    ...                 [0.20, 0.10, 0.20]], 'a=a0,a1;b=b0..b2')                        # doctest: +SKIP
+    >>> proba                                                                           # doctest: +SKIP
+    a\b    b0    b1   b2
+     a0  0.15  0.25  0.1
+     a1   0.2   0.1  0.2
+    >>> choice(p=proba, axes='draw=d0..d5')                                             # doctest: +SKIP
+    draw\axis   a   b
+           d0  a1  b2
+           d1  a1  b1
+           d2  a0  b1
+           d3  a0  b0
+           d4  a1  b2
+           d5  a0  b1
     """
     axes = AxisCollection(axes)
     if isinstance(p, LArray):
         if choices is not None:
             raise ValueError("choices argument cannot be used when p argument is an LArray")
 
         if p.ndim > 1:
-            raise ValueError("using an larray with ndim > 1 for p is currently not supported")
-        choices = p.axes[0].labels
-        p = p.data
+            flat_p = p.data.reshape(-1)
+            flat_indices = choice(p.size, axes=axes, replace=replace, p=flat_p)
+            return axes_flat_lookup(p.axes, flat_indices)
+        else:
+            choices = p.axes[0].labels
+            p = p.data
     if choices is None:
         raise ValueError("choices argument must be provided unless p is an LArray")
     return LArray(np.random.choice(choices, axes.shape, replace, p), axes, meta=meta)
+
+
+# XXX: make this into a method or property on AxisCollection? axes.flat_labels[flat_indices]?
+def axes_flat_lookup(axes, flat_indices):
```

Comment:
at least move this function to core/axis.py or make it a private method of AxisCollection.

Reply:
ok